### PR TITLE
fix case_column_or_null with nullable when conditions

### LINKED ISSUE #13885: Incorrect CASE WHEN + ELSE NULL behavior
### Describe the bug

some queries containing `case when .. ELSE NULL end` produces incorrect results.

### To Reproduce

preparing test table:
```sql
$ datafusion-cli
DataFusion CLI v43.0.0
> create table t1(s string);
0 row(s) fetched.
Elapsed 0.003 seconds.

> insert t1 (s) values ('aaa'), ('bbb');
+-------+
| count |
+-------+
| 2     |
+-------+
1 row(s) fetched.
Elapsed 0.002 seconds.

> select * from t1;
+-----+
| s   |
+-----+
| aaa |
| bbb |
+-----+
2 row(s) fetched.
Elapsed 0.001 seconds.
```

query:
```sql

> select case when (NULL and (s = 'aaa')) then 'unreachable!' else null end from t1;
+-------------------------------------------------------------------------------+
| CASE WHEN NULL AND t1.s = Utf8("aaa") THEN Utf8("unreachable!") ELSE NULL END |
+-------------------------------------------------------------------------------+
|                                                                               |
|                                                                               |
+-------------------------------------------------------------------------------+
2 row(s) fetched.
Elapsed 0.002 seconds.

> select case when (NULL and (s = 'aaa')) then s else 'always here!' end from t1;
+-------------------------------------------------------------------------------+
| CASE WHEN NULL AND t1.s = Utf8("aaa") THEN t1.s ELSE Utf8("always here!") END |
+-------------------------------------------------------------------------------+
| always here!                                                                  |
| always here!                                                                  |
+-------------------------------------------------------------------------------+
2 row(s) fetched.
Elapsed 0.001 seconds.

> select case when (NULL and (s = 'aaa')) then s else null end from t1;
+---------------------------------------------------------------+
| CASE WHEN NULL AND t1.s = Utf8("aaa") THEN t1.s ELSE NULL END |
+---------------------------------------------------------------+
| aaa                                                           | <- BUGGY, should be NULL
|                                                               |
+---------------------------------------------------------------+
2 row(s) fetched.
Elapsed 0.001 seconds.
```

### Expected behavior

```sql
select case when (NULL and (s = 'aaa')) then s else null end from t1;
```

the above query should always output nulls, since the only WHEN condition is always null (not true)

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by richox**
## Which issue does this PR close?

Closes #13885 .

### CODE REVIEW

File: datafusion/sqllogictest/test_files/case.slt [hunk 1]
```diff
@@ -50,7 +50,7 @@ NULL
 6
 NULL
 NULL
-7
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/case.slt:53]: this last row has `a IS NULL`,`b = 7`

So the idea is that a > 2 is not true when `a IS NULL` so should fall through to the `ELSE null` clause

File: datafusion/sqllogictest/test_files/case.slt [hunk 2]
```diff
@@ -61,7 +61,7 @@ NULL
...
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/case.slt:64]: sam thing here, `NULL > 2` is not `true` (it isn't false either, it is NULL) so this should not take the value of `b`

File: datafusion/sqllogictest/test_files/case.slt [hunk 3]
```diff
@@ -50,7 +50,7 @@ NULL
...
+NULL
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/case.slt:53]: Can you also please add a test to this file where there are no nulls (to ensure the codepath where `null_count = 0` is also covered)

**Review by alamb — COMMENTED** [review]
Thank you @richox  -- this code and fix makes sense to me. I think this PR just needs one more test and this PR will be good to go

File: datafusion/sqllogictest/test_files/case.slt (hunk 3 above)
Comment (richox) [reply datafusion/sqllogictest/test_files/case.slt:53]: done

**Review by alamb — APPROVED** [review]
Thank you @richox  🙏

**Review by xudong963 — APPROVED** [review]

**Comment by alamb** [comment]
Thanks @richox and @xudong963 🙏

**Pull request merged**